src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs

```diff
@@ -149,5 +165,57 @@ internal static class SemanticClassificationUtilities
                 throw ExceptionUtilities.Unreachable;
             }
         }
+
+        private static async Task AddSemanticClassificationsAsync(
+            Document document,
+            TextSpan textSpan,
+            IClassificationService classificationService,
+            List<ClassifiedSpan> classifiedSpans,
+            CancellationToken cancellationToken)
+        {
+            var workspace = document.Project.Solution.Workspace;
+            var fullyLoadedStateTask = s_workspaceToFullyLoadedStateTask.GetValue(
+                workspace, w =>
+                {
+                    var workspaceLoadedService = workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+                    return workspaceLoadedService.WaitUntilFullyLoadedAsync(CancellationToken.None);
+                });
+
+            // If we're not fully loaded try to read from the cache instead so that classifications appear up to date.
+            // New code will not be semantically classified, but will eventually when the project fully loads.
+            var isFullyLoaded = fullyLoadedStateTask.IsCompleted;
+            if (await TryAddSemanticClassificationsFromCacheAsync(document, textSpan, classifiedSpans, isFullyLoaded, cancellationToken).ConfigureAwait(false))
+                return;
+
+            await classificationService.AddSemanticClassificationsAsync(
+                document, textSpan, classifiedSpans, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<bool> TryAddSemanticClassificationsFromCacheAsync(
+            Document document,
+            TextSpan textSpan,
+            List<ClassifiedSpan> classifiedSpans,
+            bool isFullyLoaded,
+            CancellationToken cancellationToken)
+        {
+            // Don't use the cache if we're fully loaded.  We should just compute values normally.
+            if (isFullyLoaded)
+                return false;
+
+            var semanticCacheService = document.Project.Solution.Workspace.Services.GetService<ISemanticClassificationCacheService>();
+            if (semanticCacheService == null)
+                return false;
+
+            var checksums = await document.State.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+            var checksum = checksums.Text;
+
+            var result = await semanticCacheService.GetCachedSemanticClassificationsAsync(
+                (DocumentKey)document, textSpan, checksum, cancellationToken).ConfigureAwait(false);
+            if (result.IsDefaultOrEmpty)
+                return false;
```

Comment:
If a file actually has no semantic classifications in the requested span, do we want to be falling back here? This feels like we're mixing the "empty" vs "not cached" distinction and that often leads to sadness, at least in the general sense.

Reply:
i see your point.  let me revise to have a 'null/empty' semantics.

Reply:
dine.